File: src/main/java/ir/reservs/reservs/ui/main/search/SearchFragment.java
package ir.reservs.reservs.ui.main.search;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.constraintlayout.widget.ConstraintLayout;

import com.google.android.material.snackbar.Snackbar;

import javax.inject.Inject;

import ir.reservs.reservs.R;
import ir.reservs.reservs.ui.base.BaseFragment;

public class SearchFragment extends BaseFragment implements SearchContract.View {

    private TextView txtDate;
    private TextView txtCityName;
    private TextView txtLocation;

    @Inject
    SearchPresenter searchPresenter;


    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.layout_search, container, false);
    }

    @Override
    public void onResume() {
        super.onResume();
        searchPresenter.onAttach(this);
    }

    @Override
    public void setDate(int year, int month, int day) {
        txtDate.setText(year + "/" + month + "/" + day);
    }

    @Override
    public void setCity(String city) {
        txtCityName.setText(city);
    }

    @Override
    public void setLocation(String location) {
        txtLocation.setText(location);
    }

    @Override
    public void onError(String error) {
        Snackbar.make(txtDate, error, Snackbar.LENGTH_LONG).show();
    }

    @Override
    public void setup(View view) {
        getFragmentComponent().inject(this);
        txtCityName = view.findViewById(R.id.txtCityName);
        txtDate = view.findViewById(R.id.txtDate);
        txtLocation = view.findViewById(R.id.txtLocation);
        ConstraintLayout constraintLayoutDate, constraintLayoutCity, constraintLayoutLocation;
        constraintLayoutDate = view.findViewById(R.id.constraintLayoutLocation);
        constraintLayoutCity = view.findViewById(R.id.constraintLayoutCity);
        constraintLayoutLocation = view.findViewById(R.id.constraintLayoutLocation);
        constraintLayoutDate.setOnClickListener((v) -> searchPresenter.getDate());
        constraintLayoutCity.setOnClickListener((v) -> searchPresenter.getCity());
        constraintLayoutLocation.setOnClickListener((v) -> searchPresenter.getLocation());
    }

    @Override
    public void onError(int resId) {
        onError(getString(resId));
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (searchPresenter != null) {
            searchPresenter.onDetach();
            searchPresenter = null;
        }

    }
}

File: src/main/java/ir/reservs/reservs/ui/main/search/SearchFragment.kt
package ir.reservs.reservs.ui.main.search

import android.app.AlertDialog
import android.content.DialogInterface
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import androidx.fragment.app.FragmentManager
import androidx.navigation.fragment.findNavController
import ir.reservs.reservs.R
import ir.reservs.reservs.model.City
import ir.reservs.reservs.model.Salon
import ir.reservs.reservs.ui.base.BaseFragment
import ir.reservs.reservs.ui.dialog.city.SelectCityFragment
import ir.reservs.reservs.ui.dialog.city.SelectCityFragment.SelectCityListener
import ir.reservs.reservs.ui.dialog.salon.SelectSalonFragment
import ir.reservs.reservs.ui.dialog.salon.SelectSalonFragment.SelectSalonListener
import kotlinx.android.synthetic.main.layout_search.*
import javax.inject.Inject

class SearchFragment : BaseFragment(), SearchContract.View,
        SelectSalonListener, SelectCityListener {

    var searchPresenter: SearchPresenter? = null
        @Inject set

    var dialog: AlertDialog? = null

    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.layout_search, container, false)
    }

    override fun setup(view: View) {
        fragmentComponent?.inject(this)
        searchPresenter?.onAttach(this)
        constraintLayoutCity.setOnClickListener { searchPresenter?.getCity() }
        constraintLayoutSalon.setOnClickListener { searchPresenter?.getSalon() }
        constraintLayoutDate.setOnClickListener { searchPresenter?.getDate() }
        view.findViewById<Button>(R.id.btnSearch).setOnClickListener { searchPresenter?.search() }

        dialog = AlertDialog.Builder(requireContext())
                .setView(View.inflate(context, R.layout.dialog_error, null))
                .setPositiveButton(R.string.ok_btn) { dialogInterface: DialogInterface, _: Int ->
                    dialogInterface.dismiss()
                }.create()
    }

    override fun openCityDialog() {
        SelectCityFragment(this).show(getFManager(), "")
    }

    override fun openSalonDialog(cityId: Int) {
        SelectSalonFragment(this, cityId).show(getFManager(), "")
    }

    override fun displayDate(date: String) {
        txtDate.text = date
    }

    override fun displayCity(city: City) {
        txtCityName.text = city.name
    }

    override fun onSelect(salon: Salon) {
        txtLocation.text = salon.title
        searchPresenter?.setSalon(salon)
    }

    override fun onSelect(city: City) {
        txtCityName.text = city.name
        searchPresenter?.setCity(city)
    }

    override fun onError() {
        dialog?.show()
    }

    override fun getFManager(): FragmentManager {
        return childFragmentManager;
    }

    override fun goToTimesFragment(salon: Salon, date: String) {
        val b = Bundle()
        b.putParcelable("salon", salon)
        b.putString("date", date)
        findNavController().navigate(R.id.fromSearchToTimes, b)
    }

    override fun onDestroy() {
        searchPresenter?.onDetach()
        searchPresenter = null
        dialog?.dismiss()
        dialog = null
        super.onDestroy()
    }

}
